p2p/src/sync/mod.rs

```diff
@@ -89,445 +97,485 @@ where
     T::SyncingMessagingHandle: SyncingMessagingService<T>,
     T::PeerId: 'static,
 {
+    /// Creates a new sync manager instance.
     pub fn new(
         chain_config: Arc<ChainConfig>,
         p2p_config: Arc<P2pConfig>,
-        handle: T::SyncingMessagingHandle,
+        messaging_handle: T::SyncingMessagingHandle,
         chainstate_handle: subsystem::Handle<Box<dyn chainstate_interface::ChainstateInterface>>,
-        rx_sync: mpsc::UnboundedReceiver<SyncControlEvent<T>>,
-        tx_peer_manager: mpsc::UnboundedSender<PeerManagerEvent<T>>,
+        peer_event_receiver: mpsc::UnboundedReceiver<SyncControlEvent<T>>,
+        peer_manager_sender: mpsc::UnboundedSender<PeerManagerEvent<T>>,
     ) -> Self {
         Self {
-            chain_config,
-            _p2p_config: p2p_config,
-            peer_sync_handle: handle,
-            rx_sync,
-            tx_peer_manager,
+            _chain_config: chain_config,
+            p2p_config,
+            messaging_handle,
+            peer_event_receiver,
+            peer_manager_sender,
             chainstate_handle,
             peers: Default::default(),
+            blocks_queue: Default::default(),
         }
     }
 
-    /// Get mutable reference to the handle
-    pub fn handle_mut(&mut self) -> &mut T::SyncingMessagingHandle {
-        // TODO: get rid of this function as it's used only in tests; perhaps a better way to do this is by
-        // creating p2p objects and make them communicate together instead of having access to internal
-        // private parts of the sync manager
-        &mut self.peer_sync_handle
-    }
+    /// Runs the sync manager event loop.
+    pub async fn run(&mut self) -> Result<Void> {
+        log::info!("Starting SyncManager");
 
-    /// Register peer to the `SyncManager`
-    pub async fn register_peer(&mut self, peer_id: T::PeerId) -> crate::Result<()> {
-        ensure!(
-            !self.peers.contains_key(&peer_id),
-            P2pError::PeerError(PeerError::PeerAlreadyExists),
-        );
+        let mut new_tip_receiver = self.subscribe_to_new_tip().await?;
+        loop {
+            tokio::select! {
+                event = self.messaging_handle.poll_next() => match event? {
+                    SyncingEvent::Message { peer, message } => {
+                        let res = self.handle_message(peer, message).await;
+                        self.handle_result(peer, res).await?;
+                    },
+                    SyncingEvent::Announcement{ peer, announcement } => {
+                        let res = self.handle_announcement(peer, announcement).await;
```

Comment:
Why do we even need announcements? Why not just use the headers message?

Reply:
There is a special logic for the announcements in the bitcoin code even though there is no special message, so I decided to keep the separation that we already had.